gossip/tx_scrambler.go

```diff
@@ -0,0 +1,128 @@
+package gossip
+
+import (
+	"bytes"
+	"github.com/ethereum/go-ethereum/common"
+	"slices"
+)
+
+type scramblerEntry struct {
+	hash   common.Hash
+	sender common.Address
+	nonce  uint64
+	gas    uint64
+}
+
+// getExecutionOrder first removes any entries with duplicate hashes, then sorts the list by XORed hashes.
+// Furthermore, if there are entries with same sender, these entries are sorted by their nonce (lower comes first).
+func getExecutionOrder(entries []*scramblerEntry) []*scramblerEntry {
+	uniqueList, salt, hasDuplicateAddresses := analyseEntryList(entries)
+	scrambleTransactions(uniqueList, salt)
+
+	// do we need to sort more?
+	if hasDuplicateAddresses {
+		sortTransactionsWithSameSender(uniqueList)
+	}
+
+	return uniqueList
+}
+
+// sortTransactionsWithSameSender finds any duplicate senders and sorts their transactions by nonce ascending.
+func sortTransactionsWithSameSender(entries []*scramblerEntry) {
+	senderNonceOrder := slices.Clone(entries)
+	// sort copied slice so that it has all txs from same address together + sorted by nonce ascending
+	slices.SortFunc(senderNonceOrder, func(a, b *scramblerEntry) int {
+		cmp := a.sender.Cmp(b.sender)
+		if cmp != 0 {
+			return cmp
+		}
+		if a.nonce > b.nonce {
+			return 1
+		}
+		if a.nonce < b.nonce {
+			return -1
+		}
+		// if nonce is equal, sort by gas
+		if a.gas > b.gas {
+			return -1
+		} else {
+			// if both nonce and gas is equal, order between those do not matter
+			return 1
+		}
+	})
```

Comment:
To guarantee a total order, you need to consider the case where `a.sender == b.sender && a.nonce == b.nonce && a.gas == b.gas`; in this case, you should use the transaction hash as a tie breaker.